rust-symcrypt/src/hash.rs

```diff
@@ -1,59 +1,98 @@
+/* Hashing functions. For further documentation please refer to symcrypt.h */
+
 use core::ffi::c_void;
 use std::mem;
+use std::pin::Pin;
 use std::ptr;
 use symcrypt_sys;
 
 pub const SHA256_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA256_RESULT_SIZE as usize;
 pub const SHA384_RESULT_SIZE: usize = symcrypt_sys::SYMCRYPT_SHA384_RESULT_SIZE as usize;
 
-pub trait Hash {
+/// Generic trait for stateful hashing
+///
+/// append() appends to be hashed data to the state, this operation can be done multiple times.
+///
+/// result() returns the result of the hash. Once result() is called, the lifetime of the ShaXXXState is finished.
+/// To perform other stateful hash operation you must create a new hash object via ShaXXXState::new()
+///
+/// copy() creates a copy of the current ShaXXXState
+pub trait HashState {
     type Result;
 
     fn append(&mut self, data: &[u8]);
-    fn result(&mut self) -> Self::Result;
+    fn result(self) -> Self::Result;
+    fn copy(&self) -> Box<Self>;
 }
 
+/// Sha256State needs to have a heap allocated inner state that is Pin<Box<T>> since the memory address of Self is moved around when implementing
+/// HashState Result field.
+///
+/// SymCrypt expects the address for its structs to stay static through the structs lifetime to guarantee that structs are not memcpy'd as
+/// doing so would lead to use-after-free and inconsistent states.
 pub struct Sha256State {
-    state: symcrypt_sys::SYMCRYPT_SHA256_STATE,
+    state: Pin<Box<Sha256InnerState>>,
 }
 
+struct Sha256InnerState(symcrypt_sys::SYMCRYPT_SHA256_STATE);
+
+/// Creates a new instance of Sha256State, this must be called before other HashState functions can be called
 impl Sha256State {
     pub fn new() -> Self {
         let mut instance = Sha256State {
-            state: symcrypt_sys::SYMCRYPT_SHA256_STATE::default(),
+            state: Box::pin(Sha256InnerState(
+                symcrypt_sys::SYMCRYPT_SHA256_STATE::default(),
+            )),
```

Comment:
```suggestion
            state: Box::pin(Sha256InnerState::default()),
```

If you stick `#[derive(Default)]` on the struct definition you can make this a little simpler.

Reply:
I thought that this would work easily but looks like Rust cant pick up the `#[derive(Default)]` for `symcrypt_sys::SYMCRYPT_SHA256_STATE`, error : 

`the trait bound `Pin<Box<Sha256InnerState>>: Default` is not satisfied
the trait `Default` is not implemented for `Pin<Box<Sha256InnerState>>`

Even after I added `#[derive(Default)]` for both `Sha256State` and `Sha256InnerState`

for now I will keep how I have it

Reply:
👍 

After suggesting that I was playing around a little and realized `Pin` doesn't impl `Default`, so the suggestion doesn't quite work :)